model/container_task_queue.go

```diff
@@ -0,0 +1,150 @@
+package model
+
+import (
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+// ContainerTaskQueue represents an iterator that represents an ordered queue of
+// container tasks that are ready be allocated a container.
+type ContainerTaskQueue struct {
+	queue    []task.Task
+	position int
+}
+
+// Next returns the next task that's ready for container allocation. It will
+// return a nil task once there are no tasks remaining in the queue.
+func (q *ContainerTaskQueue) Next() (*task.Task, error) {
+	if q.queue == nil {
+		if err := q.populate(); err != nil {
+			return nil, errors.Wrap(err, "initial population of container task queue")
+		}
+	}
+	if q.position >= len(q.queue) {
+		return nil, nil
+	}
+
+	next := q.queue[q.position]
+
+	q.position++
+
+	return &next, nil
+}
+
+func (q *ContainerTaskQueue) populate() error {
+	candidates, err := task.FindNeedsContainerAllocation()
+	if err != nil {
+		return errors.Wrap(err, "finding candidate container tasks for allocation")
+	}
+
+	readyForAllocation, err := q.filterByProjectRefSettings(candidates)
+	if err != nil {
+		return errors.Wrap(err, "filtering candidate container tasks for allocation by project ref settings")
+	}
+
+	q.queue = readyForAllocation
+
+	return nil
+}
+
+func (q *ContainerTaskQueue) filterByProjectRefSettings(tasks []task.Task) ([]task.Task, error) {
+	projRefs, err := q.getProjectRefs(tasks)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting project refs")
+	}
+
+	var readyForAllocation []task.Task
+	for _, t := range tasks {
+		ref, ok := projRefs[t.Project]
+		if !ok {
+			grip.Warning(message.Fields{
+				"message": "skipping task that is a candidate for allocation because did not find the project associated with it",
+				"outcome": "skipping",
+				"task":    t.Id,
+				"project": t.Project,
+				"context": "container task queue",
+			})
+			continue
+		}
+
+		if !ref.IsEnabled() {
+			// GitHub PR tasks are still allowed to run for disabled hidden
+			// projects.
+			if t.Requester == evergreen.GithubPRRequester && ref.IsHidden() {
+				grip.Info(message.Fields{
```

Comment:
Should this be a debug message?

Reply:
Probably yeah. I changed it and all the other logs to debug since they're mostly for tracing purposes.